### PR TITLE
Two StandardOutput=truncate:file improvements

### PR DESCRIPTION
**Pull request by lucaswerkmeister**
One test fix (see commit message for details) and one documentation addition.

### CODE REVIEW

File: man/systemd.exec.xml [hunk 1]
```diff
@@ -2425,8 +2425,12 @@ SystemCallErrorNumber=EPERM</programlisting>
...
 
         <para><option>truncate:<replaceable>path</replaceable></option> is similar to
-        <option>file:<replaceable>path</replaceable></option> above, but it truncates the file when opening it.
-        </para>
+        <option>file:<replaceable>path</replaceable></option> above, but it truncates the file when opening
+        it. For units with multiple command lines, e.g. <varname>Type=oneshot</varname> services with
+        multiple <varname>ExecStart=</varname>, or services with <varname>ExecCondition=</varname>,
+        <varname>ExecStartPre=</varname> or <varname>ExecStartPost=</varname>, the output file is reopened
+        and therefore re-truncated for each command line. Any <varname>ExecReload=</varname> will likewise
+        truncate the file when run.</para>
```
Comment (lucaswerkmeister) [inline man/systemd.exec.xml]: I tested the `ExecReload=` part with the following service:

```ini
[Service]
ExecStart=sh -c 'echo started; sleep 10s; echo stopped'
ExecReload=true
StandardOutput=truncate:/tmp/out
```

If the service is left to complete on its own, `/tmp/out` will contain “started” and “stopped”; if you `systemctl reload` the service during the 10s sleep, then `/tmp/out` will only contain “stopped” at the end. (That is, the running main service can continue to write to its stdout fd at its current offset and the output will end up in the file, the reload just discards all earlier output.)

Comment (lucaswerkmeister) [reply man/systemd.exec.xml]: > (That is, the running main service can continue to write to its stdout fd at its current offset and the output will end up in the file, the reload just discards all earlier output.)

Wait, no, the output from `cat /tmp/out` confused me. `hexdump` makes it clearer:

```console
$ hexdump -C /tmp/out
00000000  00 00 00 00 00 00 00 00  73 74 6f 70 70 65 64 0a  |........stopped.|
00000010
```

So the main process’ stdout offset isn’t adjusted, the file is just filled with null bytes. (I haven’t tested yet whether those null bytes are always written to disk or whether, for offsets above the fs block size, the file becomes sparse.)

Maybe the documentation’s advice should be “do not use truncate mode with services that have more than one command line; it’s unlikely to do what you want.”

**Comment by boucman** [comment]
I wonder if it's a bug rather than a documentation problem...

File: man/systemd.exec.xml (hunk 1 above)
Comment (poettering) [reply man/systemd.exec.xml]: This is expected of truncation: the truncation doesn#t change the file pointer of the processes that already have the file open, and thus they will point to an offset after the file length. THe instant they write something to it, the file will be extended though, filling what comes between with zeroes. (well, actually the file becomes a sparse file modulo the block size, which is effectively the same)

Comment (poettering) [reply man/systemd.exec.xml]: might make sense to add a brief comment about this to the man page. maybe: "Note that truncation does not reset the file write pointers of the processes that already have the file open. They will continue writing to the offset they are pointed to, which might have the effect to insert space (i.e. NUL bytes) in between the write pointers of the existing processes and the new processes, for example when ExectStart= and ExecReload= are run concurrently. Thus, typically `truncate:` is only useful if only one process of the service runs at a time, i.e. only a single ExecStart= line is defined in the service, without ExecStartPost=, ExecReload=, ExecStop= and similar."

or something along those lines

**Review by poettering — CHANGES_REQUESTED** [review]

File: man/systemd.exec.xml (hunk 1 above)
Comment (lucaswerkmeister) [reply man/systemd.exec.xml]: I added something along those lines and put you in Co-authored-by, hope that’s okay :)

**Pull request merged**